### PR TITLE
fix - destroy model hangs for caas model with detached volumes

### PR DESCRIPTION
**Pull request by ycliuhw**
## Description of change

fix destroy model hangs if there are volumes in the model.

## QA steps

1. prepare a caas model;
2. deploy stateful caas charm(mariadb-k8s for example);
3. remove application without removing storage: `juju remove-application mariadb-k8s`
4. check the storage are detached after application was removed: `juju storage [--filesystem]`
5. destroy model with `--destroy-storage` enabled: `juju destroy-model t1 --debug -y --destroy-storage`

## Documentation changes

No doc change because it's a bug fix.

### CODE REVIEW

**Comment by ycliuhw** [comment]
[model destroy hangs, remove application fails](https://trello.com/c/pC3kcO1c/110-model-destroy-hangs-remove-application-fails)

**Comment by ycliuhw** [comment]
!!build!!

**Comment by ycliuhw** [comment]
[remove application can cause NPE in storage provisioner worker](https://trello.com/c/wX6aveto/501-remove-application-can-cause-npe-in-storage-provisioner-worker)

File: worker/storageprovisioner/filesystem_events.go [hunk 1]
```diff
@@ -77,7 +77,7 @@ func filesystemAttachmentsChanged(ctx *context, watcherIds []watcher.MachineStor
 	if len(dead) != 0 {
 		// We should not see dead filesystem attachments;
 		// attachments go directly from Dying to removed.
-		logger.Warningf("unexpected dead filesystem attachments: %v", dead)
+		logger.Warningf("dead filesystem attachments: %v", dead)
```
Comment (wallyworld) [inline worker/storageprovisioner/filesystem_events.go]: we want to keep the "unexpected"

File: worker/storageprovisioner/storageprovisioner.go [hunk 2]
```diff
@@ -531,3 +530,7 @@ type context struct {
 	// machine.
 	managedFilesystemSource storage.FilesystemSource
 }
+
+func (c *context) IsApplicationKind() bool {
```
Comment (wallyworld) [inline worker/storageprovisioner/storageprovisioner.go]: Does this need to be exported? 
If so, it needs a comment, if not make it private.
Also, I think there's another place we can use this helper, where we add the watchers

File: worker/storageprovisioner/volume_events.go [hunk 3]
```diff
@@ -182,7 +182,7 @@ func volumeAttachmentsChanged(ctx *context, watcherIds []watcher.MachineStorageI
 	if len(dead) != 0 {
 		// We should not see dead volume attachments;
 		// attachments go directly from Dying to removed.
-		logger.Warningf("unexpected dead volume attachments: %v", dead)
+		logger.Warningf("dead volume attachments: %v", dead)
```
Comment (wallyworld) [inline worker/storageprovisioner/volume_events.go]: ditto

File: worker/storageprovisioner/volume_events.go [hunk 4]
```diff
@@ -214,6 +214,9 @@ func volumeAttachmentsChanged(ctx *context, watcherIds []watcher.MachineStorageI
 // processDyingVolumes processes the VolumeResults for Dying volumes,
 // removing them from provisioning-pending as necessary.
 func processDyingVolumes(ctx *context, tags []names.Tag) error {
+	if ctx.IsApplicationKind() {
```
Comment (wallyworld) [inline worker/storageprovisioner/volume_events.go]: Maybe add a comment why we can ignore the volumes, here and in alive()

File: caas/kubernetes/provider/storage.go [hunk 5]
```diff
@@ -228,9 +229,13 @@ func (v *volumeSource) DestroyVolumes(ctx context.ProviderCallContext, volIds []
 	logger.Debugf("destroy k8s volumes: %v", volIds)
 	pVolumes := v.client.CoreV1().PersistentVolumes()
 	return foreachVolume(volIds, func(volumeId string) error {
-		return pVolumes.Delete(volumeId, &v1.DeleteOptions{
-			PropagationPolicy: &defaultPropagationPolicy,
-		})
+		if err := pVolumes.Delete(
```
Comment (wallyworld) [inline caas/kubernetes/provider/storage.go]: really should add or update a test for this

**Comment by ycliuhw** [comment]
$$merge$$

**Pull request merged**